proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+---
+simd: 'xxxx'
+title: Asynchronous Epoch Reward Distribution 
+authors:
+  - Haoran Yi (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-03-02
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+A new way to distribute epoch rewards asynchronously is proposed to mitigate the current performance issues with the current synchronous distribution method.
+
+## Motivation
+
+The distribution of epoch rewards at the beginning of an epoch becomes a
+significant bottleneck due to the rising number of stake accounts and voting
+nodes on the network.
+
+To address this bottleneck, we propose a new approach for distributing the
+epoch rewards asynchronously. Instead of distributing rewards at the start of
+the epoch synchronously, we propose an asynchronous distribution method that spreads the
+rewards payout over multiple blocks.
+
+
+## Detailed Design
+
+The major bottleneck is paying out rewards to stake accounts, which is about
+550K, while rewards for vote accounts are around 15K. Given the relatively
+small number of vote rewards, it make sense to keep vote rewards payout the
+same. They can still be paid out synchronously and efficiently at the first
+block of the epoch boundary. This reduce the impact on rewards for vote account
+and also simplify the over changes, and let us focus on solving the primary
+bottleneck - Stake Rewards. Stake rewards are going to be paid out
+asynchronously over multiple blocks.
+
+For async stake rewards payout, we will separate the computation of rewards
+from the actual distribution of rewards at the epoch boundary by dividing the
+process into two distinct phases: (1) rewards calculation, and (2) rewards
+distribution.
+
+
+### Rewards Calculation
+
+In contrast to the current method of computing epoch rewards, which involves
+directly computing rewards at the first slot of the epoch boundary in the bank
+runtime, the new approach involves moving the reward computation to a
+background service known as the `EpochRewardCalculationService`. This service
+starts automatically when a validator starts and listens to a designated
+channel for any incoming rewards calculation requests, performing the necessary
+rewards calculations accordingly.
+
+When a block first crosses the epoch boundary, its bank will send a request to
+the `EpochRewardCalculationService` to start the process of epoch reward
+computation. This marks the start of the reward computation phase.
+
+However, multiple rewards computation requests can be generated when there are
+forks at the epoch boundaries.
```

Comment:
> 

Interesting!

But I don't think this will work with StakeCache. I believe that StakeCache is shared for all the banks in the same fork. That means when we start reward computation when the epoch start bank is rooted. The StakeCache has already moved on. 
To make this work, we will need to reload all stake accounts from accounts_db then.

Reply:
> an we send the request to EpochRewardCalculationService when the bank is rooted instead?

No, because roots are local to a validator and not deterministic across the entire cluster

Reply:
Yeah. We can omit the implementation details from the proposal.